espressostreamer/espresso_streamer.go

```diff
@@ -0,0 +1,306 @@
+	espressoClients               []espressoClient.Client
+	nextHotshotBlockNum           uint64
+	currentMessagePos             uint64
+	namespace                     uint64
+	confirmedMessagePos           uint64
+	retryTime                     time.Duration
+	pollingHotshotPollingInterval time.Duration
+	messageWithMetadataAndPos     []*MessageWithMetadataAndPos
+	espressoTEEVerifierABI        *abi.ABI
+	espressoTEEVerifierAddr       common.Address
+}
+
+func NewEspressoStreamer(namespace uint64, hotshotUrls []string,
+	nextHotshotBlockNum uint64,
+	retryTime time.Duration,
+	pollingHotshotPollingInterval time.Duration,
+	parentChainNodeUrl string,
+	headerReaderConfig headerreader.Config,
+	espressoTEEVerifierAddress common.Address,
+) *EspressoStreamer {
+	var espressoClients []espressoClient.Client
+	for _, url := range hotshotUrls {
+		client := espressoClient.NewClient(url)
+		if client == nil {
+			// we should be able to initialize
+			// all the clients
+			log.Crit("Failed to create espresso client", "url", url)
+			return nil
+		}
+		espressoClients = append(espressoClients, *client)
+	}
+
+	espressoTEEVerifierAbi, err := bridgegen.IEspressoTEEVerifierMetaData.GetAbi()
+	if err != nil {
+		log.Crit("Unable to find EspressoTEEVerifier ABI")
+	}
+	l1Client, err := ethclient.Dial(parentChainNodeUrl)
+	if err != nil {
+		log.Crit("Failed to create l1 client", "url", parentChainNodeUrl)
+		return nil
+	}
+
+	arbSys, err := precompilesgen.NewArbSys(types.ArbSysAddress, l1Client)
+	if err != nil {
+		log.Crit("Failed to create arbsys", "err", err)
+		return nil
+	}
+
+	// we initialze a l1 reader that will poll for header every 60 seconds
+	l1Reader, err := headerreader.New(context.Background(), l1Client, func() *headerreader.Config {
+		return &headerReaderConfig
+	}, arbSys)
+
+	if err != nil {
+		log.Crit("Failed to create l1 reader", "err", err)
+		return nil
+	}
+	return &EspressoStreamer{
+		espressoClients:               espressoClients,
+		nextHotshotBlockNum:           nextHotshotBlockNum,
+		retryTime:                     retryTime,
+		pollingHotshotPollingInterval: pollingHotshotPollingInterval,
+		namespace:                     namespace,
+		l1Reader:                      l1Reader,
+		espressoTEEVerifierABI:        espressoTEEVerifierAbi,
+		espressoTEEVerifierAddr:       espressoTEEVerifierAddress,
+		// Initally, we set the confirmed message pos to max uint64
+		// because we dont know the confirmed message pos yet
+		confirmedMessagePos: abi.MaxUint256.Uint64(),
+	}
+}
+
+func (s *EspressoStreamer) Refresh(ctx context.Context, fetchLastMessageAndEspressoBlock func() (uint64, uint64, error)) (bool, error) {
+	lastMessagePos, lastEspressoBlock, err := fetchLastMessageAndEspressoBlock()
+	if err != nil {
+		return false, err
+	}
+
+	// If the last message pos only increase by 1, we dont need to refresh
+	// because this is expected behavior.
+	if lastMessagePos == s.confirmedMessagePos+1 {
```

Comment:
I understand why it's a expected behavior for the last message pos to differ every time in the caff node, but what if a batch has been posted with exactly 1 new message, this would lead to us not detecting a change in the batch poster when we call refresh even if another node would posted a batch.

Reply:
It will still hold true, so lets say the last `confirmedMessagePos` was `1` then even if only one message is posted, the `lastMessagePos` returned will be 2 which will be `confirmedMessagePos` + 1